File: mobile/lib/core/network/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../config/app_config.dart';
import '../../features/auth/data/auth_repository.dart';
import 'api_exceptions.dart';

/// Low-level HTTP client. Attaches the Firebase ID token and maps API errors.
class ApiClient {
  final Dio _dio;
  final Ref _ref;

  ApiClient(this._ref, {Dio? dio})
    : _dio =
          dio ??
          Dio(
            BaseOptions(
              baseUrl: AppConfig.apiV1BaseUrl,
              connectTimeout: const Duration(seconds: 10),
              receiveTimeout: const Duration(seconds: 10),
              headers: const {'Content-Type': 'application/json'},
            ),
          ) {
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          final authRepo = _ref.read(authRepositoryProvider);
          final user = authRepo.currentUser;
          if (user != null) {
            final idToken = await user.getIdToken();
            options.headers['Authorization'] = 'Bearer $idToken';
          }
          handler.next(options);
        },
        onError: (error, handler) {
          handler.reject(
            DioException(
              requestOptions: error.requestOptions,
              response: error.response,
              type: error.type,
              error: mapDioException(error),
            ),
          );
        },
      ),
    );
  }

  /// Maps a [DioException] to a typed [ApiException] (or [ApiAuthException] for 401).
  static ApiException mapDioException(DioException error) {
    final response = error.response;
    if (response == null) {
      return ApiException(0, error.message ?? 'Network error');
    }

    final statusCode = response.statusCode ?? 0;
    var message = error.message ?? 'Request failed';

    final data = response.data;
    if (data is Map && data['detail'] != null) {
      message = data['detail'].toString();
    }

    if (statusCode == 401) {
      return ApiAuthException(message, data);
    }

    return ApiException(statusCode, message, data);
  }

  Future<Map<String, dynamic>> get(String path) async {
    try {
      final response = await _dio.get<Map<String, dynamic>>(path);
      final data = response.data;
      if (data == null) {
        throw ApiException(response.statusCode ?? 0, 'Empty response body');
      }
      return data;
    } on DioException catch (error) {
      final mapped = error.error;
      if (mapped is ApiException) throw mapped;
      throw mapDioException(error);
    }
  }

  Future<Map<String, dynamic>> patch(
    String path, {
    required Map<String, dynamic> body,
  }) async {
    try {
      final response = await _dio.patch<Map<String, dynamic>>(path, data: body);
      final data = response.data;
      if (data == null) {
        throw ApiException(response.statusCode ?? 0, 'Empty response body');
      }
      return data;
    } on DioException catch (error) {
      final mapped = error.error;
      if (mapped is ApiException) throw mapped;
      throw mapDioException(error);
    }
  }
}

final apiClientProvider = Provider<ApiClient>((ref) => ApiClient(ref));

File: mobile-rn/src/lib/api-client.ts
/**
 * Low-level HTTP client. Attaches the Firebase ID token and maps API errors.
 * Parity with Flutter `mobile/lib/core/network/api_client.dart`.
 *
 * Only feature `api.ts` files may call this — see docs/best-practices.md §1.
 */

import { ApiAuthError, ApiError } from './api-errors';

/** Matches Flutter's Dio connect/receive timeouts. */
const TIMEOUT_MS = 10_000;

type HttpMethod = 'GET' | 'POST' | 'PATCH' | 'DELETE';

type RequestOptions = {
  method?: HttpMethod;
  body?: unknown;
  query?: Record<string, string | number | boolean | undefined>;
  /** Skip the Authorization header (e.g. unauthenticated probes). */
  anonymous?: boolean;
};

/**
 * Token provider, injected from the auth layer to avoid a circular import
 * (auth → api-client → auth). Set once at app bootstrap.
 */
type TokenProvider = () => Promise<string | null>;

let getAccessToken: TokenProvider = async () => null;

export function setAccessTokenProvider(provider: TokenProvider): void {
  getAccessToken = provider;
}

function baseUrl(): string {
  const url = process.env.EXPO_PUBLIC_API_BASE_URL;
  if (!url) {
    throw new Error(
      'EXPO_PUBLIC_API_BASE_URL is not set — copy .env.example to .env.local',
    );
  }
  return url.replace(/\/$/, '');
}

function buildUrl(path: string, query?: RequestOptions['query']): string {
  const url = `${baseUrl()}${path}`;
  if (!query) return url;

  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (value !== undefined) params.append(key, String(value));
  }
  const qs = params.toString();
  return qs ? `${url}?${qs}` : url;
}

/**
 * Maps a non-OK response to a typed error, mirroring Dio's `mapDioException`:
 * the `{ detail }` body becomes the message; 401 becomes ApiAuthError.
 */
function toApiError(status: number, body: unknown, fallback: string): ApiError {
  let message = fallback;
  if (body && typeof body === 'object' && 'detail' in body) {
    const detail = (body as { detail: unknown }).detail;
    if (detail != null) message = String(detail);
  }
  return status === 401 ? new ApiAuthError(message, body) : new ApiError(status, message, body);
}

async function request<T>(path: string, options: RequestOptions = {}): Promise<T> {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' };

  if (!options.anonymous) {
    const token = await getAccessToken();
    if (token) headers.Authorization = `Bearer ${token}`;
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);

  let response: Response;
  try {
    response = await fetch(buildUrl(path, options.query), {
      method: options.method ?? 'GET',
      headers,
      body: options.body === undefined ? undefined : JSON.stringify(options.body),
      signal: controller.signal,
    });
  } catch (error) {
    // No response at all — Flutter maps this to ApiException(0, message).
    const message = error instanceof Error ? error.message : 'Network error';
    throw new ApiError(0, message);
  } finally {
    clearTimeout(timer);
  }

  const text = await response.text();
  let data: unknown = null;
  if (text) {
    try {
      data = JSON.parse(text);
    } catch {
      data = text;
    }
  }

  if (!response.ok) {
    throw toApiError(response.status, data, `Request failed with status ${response.status}`);
  }

  return data as T;
}

export const apiClient = {
  get: <T>(path: string, query?: RequestOptions['query']) =>
    request<T>(path, { method: 'GET', query }),

  post: <T>(path: string, body: unknown) => request<T>(path, { method: 'POST', body }),

  patch: <T>(path: string, body: unknown) => request<T>(path, { method: 'PATCH', body }),

  /** DELETE returns 204 with no body — resolves to void. */
  delete: async (path: string): Promise<void> => {
    await request<unknown>(path, { method: 'DELETE' });
  },
};
